File: tests/pybind_test.cpp
// Copyright 2024 Universidad Politécnica de Madrid
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Universidad Politécnica de Madrid nor the names
//    of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

/**
 * @file pybind_test.hpp
 *
 * @brief Test class bind to python for the Dynamic Trajectory Generator.
 *
 * @author Rafael Pérez Seguí
 */

#include <Eigen/Dense>

#include <iostream>
#include <vector>

#include "_dynamic_trajectory_generator_pybind.hpp"

int main()
{
  dynamic_traj_generator::DynamicTrajectoryBind trajectory_bind;

  // Initial position of the vehicle
  const Eigen::Vector3d initial_position(0.0, 0.0, 1.0);
  const float initial_yaw = 0.0f;

  // Trajectory waypoints
  std::vector<Eigen::Vector3d> waypoints;
  waypoints.push_back(Eigen::Vector3d(1.0, 1.0, 1.0));
  waypoints.push_back(Eigen::Vector3d(1.0, -1.0, 1.0));
  waypoints.push_back(Eigen::Vector3d(-1.0, -1.0, 1.0));
  waypoints.push_back(Eigen::Vector3d(-1.0, 1.0, 1.0));
  waypoints.push_back(Eigen::Vector3d(0.0, 0.0, 1.0));

  // Generate trajectory
  trajectory_bind.setPathFacing(true);
  const float max_velocity = 2.0f;

  std::cout << "Generating trajectory..." << std::endl;
  trajectory_bind.generateTrajectory(
    initial_position,
    initial_yaw,
    waypoints,
    max_velocity);

  // Get max and min time
  std::cout << "Getting max and min time..." << std::endl;
  const float max_time = trajectory_bind.getMaxTime();
  const float min_time = trajectory_bind.getMinTime();

  // References
  Eigen::Vector3d position;
  Eigen::Vector3d velocity;
  Eigen::Vector3d acceleration;
  float yaw;

  // Evaluate trajectory
  const float dt = 0.01f;

  for (float t = min_time; t < max_time; t += dt) {
    std::tie(position, velocity, acceleration, yaw) =
      trajectory_bind.evaluateTrajectory(t);
  }

  std::cout << "Generate trajectory from: " << initial_position.transpose() << std::endl;
  std::cout << "And yaw: " << initial_yaw << std::endl;
  std::cout << "Generate trajectory with max velocity: " << max_velocity << std::endl;
  std::cout << "Generate trajectory for waypoints: " << std::endl;
  for (int i = 0; i < waypoints.size(); i++) {
    std::cout << waypoints[i].transpose() << std::endl;
  }

  std::cout << "Trajectory generated with:" << std::endl;
  std::cout << "Min time: " << min_time << std::endl;
  std::cout << "Max time: " << max_time << std::endl;

  std::cout << "Trajectory final references:" << std::endl;
  std::cout << "Position: " << position.transpose() << std::endl;
  std::cout << "Velocity: " << velocity.transpose() << std::endl;
  std::cout << "Acceleration: " << acceleration.transpose() << std::endl;
  std::cout << "Yaw: " << yaw << std::endl;

  return 0;
}

File: pybind_test/pybind_test.py
import trajectory_generator as dtb
import numpy as np

# Instance the dynamic trajectory class
trajectory_bind = dtb.DynamicTrajectory()

# Initial position of the vehicle
initial_position = np.array([0.0, 0.0, 1.0])
initial_yaw = 0.0

# Trajectory waypoints
waypoints = [
    np.array([1.0, 1.0, 1.0]),
    np.array([1.0, -1.0, 1.0]),
    np.array([-1.0, -1.0, 1.0]),
    np.array([-1.0, 1.0, 1.0]),
    np.array([0.0, 0.0, 1.0])
]

# Generate trajectory
max_velocity = 2.0
trajectory_bind.set_path_facing(True)
trajectory_bind.generate_trajectory(
    initial_position, initial_yaw,
    waypoints, max_velocity)

# Get max and min time
max_time = trajectory_bind.get_max_time()
min_time = trajectory_bind.get_min_time()

# Evaluate trajectory
time = min_time
position = np.zeros(3)
velocity = np.zeros(3)
acceleration = np.zeros(3)
yaw = 0.0

dt = 0.01
while time < max_time:
    position, velocity, acceleration, yaw = \
        trajectory_bind.evaluate_trajectory(time)
    time += dt

print("Trajectory generated from: ", initial_position)
print("And yaw: ", initial_yaw)
print("Generate trajectory with max velocity: ", max_velocity)
print("Generate trajectory for waypoints: ")
for i in range(len(waypoints)):
    print(waypoints[i])

print("Trajectory generated with:")
print("Min time: ", min_time)
print("Max time: ", max_time)

print("Trajectory final references:")
print("Position: ", position)
print("Velocity: ", velocity)
print("Acceleration: ", acceleration)
print("Yaw: ", yaw)

